bndtools.core/src/bndtools/central/Central.java

```diff
@@ -239,11 +238,21 @@ public synchronized static Workspace getWorkspace() throws Exception {
         }
     }
 
-    public synchronized static void onWorkspaceInit(Function<Workspace,Void> callback) {
-        if (workspace != null)
-            callback.run(workspace);
-        else
-            workspaceInitCallbackQueue.add(callback);
+    public static void onWorkspaceInit(Function<Workspace,Void> callback) {
+        callbacks.add(callback);
+        notifyCallbacks();
+    }
+
+    private static void notifyCallbacks() {
+        synchronized (instance) {
+            if (workspace == null)
+                return;
+        }
+
+        while (!callbacks.isEmpty()) {
+            Function<Workspace,Void> callback = callbacks.poll();
+            callback.apply(workspace);
```

Comment:
Yes. We could use a Promise to hold the workspace object. Then onWorkspaceInit would just add the callback to the Promise and naturally call it when the Promise is resolved with the workspace object. This would be better than trying to replicate the Promise logic for callback notification.